### ISSUE TITLE
Detect and warn about infinite recursion (where possible)

### ISSUE THREAD

**Issue by psybers** [boalang/compiler#14]
````sh
visitor {
   before n: Node -> visit(n);
}

f := function() { f(); };
````

Clearly these cases (and many more) can be detected statically and we should warn on such cases.

The basic analysis is to find a recursive call that is not control-dependent on any other statement and then mark it an error.  This is sound, but no-where near complete.
